src/main/java/com/uid2/admin/vertx/service/ServiceLinkService.java

```diff
@@ -118,6 +128,102 @@ private void handleServiceLinkAdd(RoutingContext rc) {
+                ResponseUtil.error(rc, 400, "json payload required but not provided");
+                return;
+            }
+            String linkId = body.getString("link_id");
+            Integer serviceId = body.getInteger("service_id");
+            Integer siteId = body.getInteger("site_id");
+            String name = body.getString("name");
+            if (siteId == null || serviceId == null) {
+                ResponseUtil.error(rc, 400, "required parameters: site_id, service_id");
+                return;
+            }
+
+            if (serviceProvider.getService(serviceId) == null) {
+                ResponseUtil.error(rc, 404, "service_id " + serviceId + " not valid");
+                return;
+            }
+
+            if (siteProvider.getSite(siteId) == null) {
+                ResponseUtil.error(rc, 404, "site_id " + siteId + " not valid");
+                return;
+            }
+
+            ServiceLink serviceLink = this.serviceLinkProvider.getAllServiceLinks()
+                    .stream().filter(s -> s.getServiceId() == serviceId && s.getSiteId() == siteId)
+                    .findFirst()
+                    .orElse(null);
+            if (serviceLink == null) {
+                ResponseUtil.error(rc, 404, "failed to find a service_link for serviceId: " + serviceId + " and site_id: " + siteId);
+                return;
+            }
+
+            if (linkId != null && !linkId.isEmpty()) {
+                serviceLink.setLinkId(linkId);
+            }
+
+            if (name != null && !name.isEmpty()) {
+                serviceLink.setName(name);
+            }
+
+            final List<ServiceLink> serviceLinks = this.serviceLinkProvider.getAllServiceLinks()
+                    .stream().sorted(Comparator.comparing(ServiceLink::getLinkId))
+                    .collect(Collectors.toList());
+
+            storeWriter.upload(serviceLinks, null);
+
+            rc.response().end(toJson(serviceLink).encodePrettily());
+        } catch (Exception e) {
+            ResponseUtil.errorInternal(rc, "Internal Server Error", e);
+        }
+    }
+
+    private void handleServiceLinkDelete(RoutingContext rc) {
+        try {
+            serviceProvider.loadContent();
+            JsonObject body = rc.body().asJsonObject();
+            if (body == null) {
+                ResponseUtil.error(rc, 400, "json payload required but not provided");
+                return;
+            }
+
+            String linkId = body.getString("link_id");
+            Integer serviceId = body.getInteger("service_id");
+
+            if (linkId == null || serviceId == null) {
+                ResponseUtil.error(rc, 400, "required parameters: link_id, service_id");
+                return;
+            }
+
+            ServiceLink serviceLink = this.serviceLinkProvider.getAllServiceLinks()
+                    .stream().filter(s -> s.getServiceId() == serviceId && s.getLinkId().equals(linkId))
+                    .findFirst()
```

Comment:
If we are allowing multiple service_links to a service, will this be enough to identify the service_link

Reply:
The linkId must be unique within a service. When authorizing a caller, we only use the LinkId and the service, so we don't want to allow duplicates. On create, we check it is unique on those 2 properties.

Reply:
Ok, makes sense